src/providers/microsoft/azure/mod.rs

```diff
@@ -412,7 +412,22 @@ impl MetadataProvider for Azure {
             bail!("unexpected empty certificates endpoint");
         }
 
-        let key = self.get_ssh_pubkey(certs_endpoint)?;
+        // The certs endpoint is being populated even when the user does not supply
+        // an SSH key when creating a VM. In this case the certificate chain is empty
+        // causing a failure when trying to access it. Check if the certificate chain
+        // is empty, meaning that an SSH key hasn't been provided, and only warn
+        // instead of fail.
+        let key = match self.get_ssh_pubkey(certs_endpoint) {
+            Ok(k) => k,
+            Err(e) => match e.root_cause().to_string().as_ref() {
+                "detected empty certificate chain" => {
+                    warn!("No SSH keys found in the certificate chain for this instance");
```

Comment:
Is this cause for warning?

Reply:
maybe not warn, but at least an info log statement?

Reply:
I'm not sure if it is cause for warning, but I chose to warn here because that's the behavior we expected when an SSH key wasn't supplied before this issue arose.

Reply:
So to make sure I understand this: before, Azure would not offer up a certificate endpoint if no SSH keys were provided, but now it always does and instead signals no SSH keys through the certificate itself?

Reply:
Basically, yes. I don't think Azure is directly signaling that no SSH key is provided through the certificate, but that's what we've found so far: that the certificate chain is empty when no SSH key is provided.

Reply:
In that case, if the new behaviour is that the certificate chain is empty when no SSH keys are present, then it should not even be considered a warning since it's completely supported to not pass an SSH key. For the same reason as the other comment, I wouldn't even `info!` it (just like other providers don't `info!` if they didn't get a key), but not strongly against. We already log something when we do write SSH keys.

`debug!()` sounds appropriate as a debugging aid if we want and could live directly in `src/providers/microsoft/crypto/mod.rs` where we `return Ok(None)` so that the calling code here can stay simple with `unwrap_or_default()` (otherwise, to log it here, we'd need to inspect the `Option` first to know if it's `None`).